SignalMessaging/ViewControllers/OWSNavigationController.m

```diff
@@ -0,0 +1,146 @@
+
+    // wasBackButtonClicked is YES if the back button was pressed but not
+    // if a back gesture was performed or if the view is popped programmatically.
+    BOOL wasBackButtonClicked = topViewController.navigationItem == item;
+    BOOL result = YES;
+    if (wasBackButtonClicked) {
+        if ([topViewController conformsToProtocol:@protocol(OWSNavigationView)]) {
+            id<OWSNavigationView> navigationView = (id<OWSNavigationView>)topViewController;
+            result = ![navigationView shouldCancelNavigationBack];
+        }
+    }
+
+    // If we're not going to cancel the pop/back, we need to call the super
+    // implementation since it has important side effects.
+    if (result) {
+        // NOTE: result might end up NO if the super implementation cancels the
+        //       the pop/back.
+        [super navigationBar:navigationBar shouldPopItem:item];
+        result =  YES;
+    }
+    return result;
+}
+
+#pragma mark - UIGestureRecognizerDelegate
+
+// We serve as the UIGestureRecognizerDelegate of the interactivePopGestureRecognizer
+// in order to cancel some "back" gestures - for example,
+// if a view has unsaved changes.
+- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
+{
+    UIViewController *topViewController = self.topViewController;
+    if ([topViewController conformsToProtocol:@protocol(OWSNavigationView)]) {
+        id<OWSNavigationView> navigationView = (id<OWSNavigationView>)topViewController;
+        return ![navigationView shouldCancelNavigationBack];
+    } else {
+        return YES;
+    }
+}
+
+
+#pragma mark - NavBarLayoutDelegate
+
+- (void)navBarCallLayoutDidChangeWithNavbar:(OWSNavigationBar *)navbar
+{
+    [self updateLayoutForNavbar:navbar];
+}
+
+- (void)updateLayoutForNavbar:(OWSNavigationBar *)navbar
+{
+    DDLogDebug(@"%@ in %s", self.logTag, __PRETTY_FUNCTION__);
+
+    if (@available(iOS 11.0, *)) {
+        if (OWSWindowManager.sharedManager.hasCall) {
+            if (UIDevice.currentDevice.isIPhoneX) {
+                // iPhoneX computes status bar height differently.
+                self.additionalSafeAreaInsets = UIEdgeInsetsMake(navbar.navbarWithoutStatusHeight + 20, 0, 0, 0);
+            } else {
+                self.additionalSafeAreaInsets
+                    = UIEdgeInsetsMake(navbar.navbarWithoutStatusHeight + CurrentAppContext().statusBarHeight, 0, 0, 0);
+            }
+        } else {
+            self.additionalSafeAreaInsets = UIEdgeInsetsZero;
+        }
+        // in iOS11 we have to ensure the navbar frame *in* layoutSubviews.
+        [navbar layoutSubviews];
+    } else {
+        // Pre iOS11 we size the navbar, and position it vertically once.
+        [navbar sizeToFit];
+
+        if (OWSWindowManager.sharedManager.hasCall) {
+            CGRect oldFrame = navbar.frame;
+            CGRect newFrame
+                = CGRectMake(oldFrame.origin.x, navbar.callBannerHeight, oldFrame.size.width, oldFrame.size.height);
```

Comment:
TIOLO but in these cases where you're mostly leaving a struct unchanged, I tend to do something like this:
```
CGRect newFrame = navbar.frame;
newFrame.height = navbar.callBannerHeight;
navbar.frame = newFrame;
```

Reply:
Better, thanks!